File: algorithms/rotate_matrix.cpp
#include "utility.h"

namespace
{
#define SIZE 5

  static void print_matrix(unsigned int matrix[SIZE][SIZE], unsigned int dimension)
  {
    for (unsigned int i = 0; i < dimension; i++)
    {
      for (unsigned int j = 0; j < dimension; j++)
      {
        std::cout << matrix[i][j] << "\t";
      }
      std::cout << std::endl;
    }
    std::cout << std::endl;
  }

  void rotate_matrix(unsigned int matrix[SIZE][SIZE], unsigned int dimension)
  {
    for (unsigned int i = 0; i < dimension / 2; i++)
    {
      unsigned int j = dimension - i - 1;
      for (unsigned int k = 0; k < j - i; k++)
      {
        unsigned int temp = matrix[i][i + k];
        matrix[i][i + k] = matrix[i + k][j];
        matrix[i + k][j] = matrix[j][j - k];
        matrix[j][j - k] = matrix[j - k][i];
        matrix[j - k][i] = temp;
      }
    }
  }
}

void test_rotate_matrix()
{
  unsigned int matrix[SIZE][SIZE] = {
    {1, 2, 3, 4, 5},
    {1, 2, 3, 4, 5},
    {1, 2, 3, 4, 5},
    {1, 2, 3, 4, 5},
    {1, 2, 3, 4, 5},
  };

  print_matrix(matrix, SIZE);
  rotate_matrix(matrix, SIZE);
  print_matrix(matrix, SIZE);
}

File: challenges/CircleCVI/rotateMatrix.js
function rotateMatrix(m) {
    const dimension = m.length;
    const iLimit = Math.floor((m.length + 1) / 2);
    const jLimit = Math.floor(m.length / 2);
    for (let i = 0; i < iLimit; i++){
        for (let j = 0; j < jLimit; j++) {
            const temp = m[j][dimension - 1 -i];
            m[j][dimension - 1 -i] = m[dimension - 1 - i][dimension - 1 -j];
            m[dimension - 1 - i][dimension - 1 -j] = m[dimension - 1 - j][i];
            m[dimension - 1 - j][i] = m[i][j];
            m[i][j] = temp;
        }
    }
}

function printMatrix(m) {
    m.forEach(row => {
        let s = ''
        row.forEach(value => s += value + ',');
        console.log(s);
    });
    console.log('');
}

[
    [
        [1, 2, 3],
        [4, 5, 6],
        [7, 8, 9]
    ],
    [
        [1, 2, 3, 4],
        [5, 6, 7, 8],
        [9, 10, 11, 12],
        [13, 14, 15, 16]
    ],
    [
        [1, 2, 3, 4, 5],
        [6, 7, 8, 9, 10],
        [11, 12, 13, 14, 15],
        [16, 17, 18, 19, 20],
        [21, 22, 23, 24, 25]
    ]
].forEach(m => {
    printMatrix(m);
    rotateMatrix(m);
    printMatrix(m);
});
